### PR TITLE
More renderer colorspace work

### PR DESCRIPTION
**Pull request by slouken**
This switches colors to float and adds the concept of colorspace to the SDL renderer, in preparation for HDR support.

### CODE REVIEW

**Comment by slouken** [comment]
@fjtrujy, can you take a look at the PS2 renderer changes in here?

File: src/render/ps2/SDL_render_ps2.c [hunk 1]
```diff
@@ -103,6 +103,27 @@ static int PixelFormatToPS2PSM(Uint32 format)
     }
 }
 
+static gs_rgbaq float_color_to_RGBAQ(const SDL_FColor *color)
+{
+    uint8_t colorR = (uint8_t)(color->r * 127.0f);
```
Comment (fjtrujy) [inline src/render/ps2/SDL_render_ps2.c:108]: I was originally wrongly doing `>> 1`, however after taking a look at the documentation it should be `255.0f` in all the channel

File: src/render/ps2/SDL_render_ps2.c [hunk 2]
```diff
@@ -103,6 +103,27 @@ static int PixelFormatToPS2PSM(Uint32 format)
...
+
+static uint64_t float_GS_SETREG_RGBAQ(const SDL_FColor *color)
+{
+    int colorR, colorG, colorB, colorA;
+
+    colorR = (int)(color->r * 127.0f);
+    colorG = (int)(color->g * 127.0f);
+    colorB = (int)(color->b * 127.0f);
+    colorA = (int)(color->a * 127.0f);
+    return GS_SETREG_RGBAQ(colorR, colorG, colorB, colorA, 0x00);
```
Comment (fjtrujy) [inline src/render/ps2/SDL_render_ps2.c:124]: Same thing here...

File: src/render/ps2/SDL_render_ps2.c (hunk 1 above)
Comment (slouken) [reply src/render/ps2/SDL_render_ps2.c:108]: Okay, I'll fix that.

**Comment by sezero** [comment]
testgeometry.c needs the following, no?
```diff
diff --git a/test/testgeometry.c b/test/testgeometry.c
index fc5ddd4..07a31a1 100644
--- a/test/testgeometry.c
+++ b/test/testgeometry.c
@@ -139,26 +139,26 @@ static void loop(void)
             a = (angle * 3.1415f) / 180.0f;
             verts[0].position.x = cx + d * SDL_cosf(a);
             verts[0].position.y = cy + d * SDL_sinf(a);
-            verts[0].color.r = 0xFF;
+            verts[0].color.r = 1.0f;
             verts[0].color.g = 0;
             verts[0].color.b = 0;
-            verts[0].color.a = 0xFF;
+            verts[0].color.a = 1.0f;
 
             a = ((angle + 120) * 3.1415f) / 180.0f;
             verts[1].position.x = cx + d * SDL_cosf(a);
             verts[1].position.y = cy + d * SDL_sinf(a);
             verts[1].color.r = 0;
-            verts[1].color.g = 0xFF;
+            verts[1].color.g = 1.0f;
             verts[1].color.b = 0;
-            verts[1].color.a = 0xFF;
+            verts[1].color.a = 1.0f;
 
             a = ((angle + 240) * 3.1415f) / 180.0f;
             verts[2].position.x = cx + d * SDL_cosf(a);
             verts[2].position.y = cy + d * SDL_sinf(a);
             verts[2].color.r = 0;
             verts[2].color.g = 0;
-            verts[2].color.b = 0xFF;
-            verts[2].color.a = 0xFF;
+            verts[2].color.b = 1.0f;
+            verts[2].color.a = 1.0f;
 
             if (use_texture) {
                 verts[0].tex_coord.x = 0.5f;
```

sdl2-compat side need updating too.

**Comment by sezero** [comment]
> sdl2-compat side need updating too.

I tried this patch (inlined below) but testgeometry doesn't look exactly the same. [compat.patch.txt](https://github.com/libsdl-org/SDL/files/14107950/compat.patch.txt)
I am missing something obvious? Will leave to you..
```diff
diff --git a/src/dynapi/SDL_dynapi_procs.h b/src/dynapi/SDL_dynapi_procs.h
index ec1877e..4b1b217 100644
--- a/src/dynapi/SDL_dynapi_procs.h
+++ b/src/dynapi/SDL_dynapi_procs.h
@@ -883,7 +883,7 @@ SDL_DYNAPI_PROC(int,SDL_JoystickSendEffect,(SDL_Joystick *a, const void *b, int
 SDL_DYNAPI_PROC(float,SDL_GameControllerGetSensorDataRate,(SDL_GameController *a, SDL_SensorType b),(a,b),return)
 SDL_DYNAPI_PROC(int,SDL_SetTextureUserData,(SDL_Texture *a, void *b),(a,b),return)
 SDL_DYNAPI_PROC(void*,SDL_GetTextureUserData,(SDL_Texture *a),(a),return)
-SDL_DYNAPI_PROC(int,SDL_RenderGeometry,(SDL_Renderer *a, SDL_Texture *b, const SDL_Vertex *c, int d, const int *e, int f),(a,b,c,d,e,f),return)
+SDL_DYNAPI_PROC(int,SDL_RenderGeometry,(SDL_Renderer *a, SDL_Texture *b, const SDL2_Vertex *c, int d, const int *e, int f),(a,b,c,d,e,f),return)
 SDL_DYNAPI_PROC(int,SDL_RenderGeometryRaw,(SDL_Renderer *a, SDL_Texture *b, const float *c, int d, const SDL_Color *e, int f, const float *g, int h, int i, const void *j, int k, int l),(a,b,c,d,e,f,g,h,i,j,k,l),return)
 SDL_DYNAPI_PROC(int,SDL_RenderSetVSync,(SDL_Renderer *a, int b),(a,b),return)
 #ifndef SDL_DYNAPI_PROC_NO_VARARGS
diff --git a/src/sdl2_compat.c b/src/sdl2_compat.c
index 107f93e..521ed14 100644
--- a/src/sdl2_compat.c
+++ b/src/sdl2_compat.c
@@ -176,6 +176,14 @@ do { \
     } \
 } while (SDL_NULL_WHILE_LOOP_CONDITION)
 
+#ifndef SDL_DISABLE_ALLOCA
+#define SDL3_stack_alloc(type, count)    (type*)alloca(sizeof(type)*(count))
+#define SDL3_stack_free(data)
+#else
+#define SDL3_stack_alloc(type, count)    (type*)SDL3_malloc(sizeof(type)*(count))
+#define SDL3_stack_free(data)            SDL3_free(data)
+#endif
+
 #include <SDL3/SDL_opengl.h>
 #include <SDL3/SDL_opengl_glext.h>
 
@@ -4237,16 +4245,40 @@ SDL_RenderCopyExF(SDL_Renderer *renderer, SDL_Texture *texture,
 }
 
 DECLSPEC int SDLCALL
-SDL_RenderGeometry(SDL_Renderer *renderer, SDL_Texture *texture, const SDL_Vertex *vertices, int num_vertices, const int *indices, int num_indices)
+SDL_RenderGeometry(SDL_Renderer *renderer, SDL_Texture *texture, const SDL2_Vertex *vertices, int num_vertices, const int *indices, int num_indices)
 {
-    const int retval = SDL3_RenderGeometry(renderer, texture, vertices, num_vertices, indices, num_indices);
-    return retval < 0 ? retval : FlushRendererIfNotBatching(renderer);
+    if (vertices) {
+        const float *xy = &vertices->position.x;
+        int xy_stride = sizeof(SDL2_Vertex);
+        const SDL_Color *color = &vertices->color;
+        int color_stride = sizeof(SDL2_Vertex);
+        const float *uv = &vertices->tex_coord.x;
+        int uv_stride = sizeof(SDL2_Vertex);
+        int size_indices = 4;
+        return SDL_RenderGeometryRaw(renderer, texture, xy, xy_stride, color, color_stride, uv, uv_stride, num_vertices, indices, num_indices, size_indices);
+    }
+    return SDL3_InvalidParamError("vertices");
 }
 
 DECLSPEC int SDLCALL
 SDL_RenderGeometryRaw(SDL_Renderer *renderer, SDL_Texture *texture, const float *xy, int xy_stride, const SDL_Color *color, int color_stride, const float *uv, int uv_stride, int num_vertices, const void *indices, int num_indices, int size_indices)
 {
-    const int retval = SDL3_RenderGeometryRaw(renderer, texture, xy, xy_stride, color, color_stride, uv, uv_stride, num_vertices, indices, num_indices, size_indices);
+    int i, retval;
+    SDL_FColor *color3 = (SDL_FColor *) SDL3_stack_alloc(SDL_FColor, num_vertices);
+    const char *color2 = (const char *) color;
+
+    for (i = 0; i < num_vertices; ++i) {
+        color3[i].r = color->r / 255.0f;
+        color3[i].g = color->g / 255.0f;
+        color3[i].b = color->b / 255.0f;
+        color3[i].a = color->a / 255.0f;
+        color2 += color_stride;
+        color = (const SDL_Color *) color2;
+    }
+
+    color_stride = sizeof(SDL_FColor);
+    retval = SDL3_RenderGeometryRaw(renderer, texture, xy, xy_stride, color3, color_stride, uv, uv_stride, num_vertices, indices, num_indices, size_indices);
+    SDL3_stack_free(color3);
     return retval < 0 ? retval : FlushRendererIfNotBatching(renderer);
 }
 
diff --git a/src/sdl2_compat.h b/src/sdl2_compat.h
index ffd96c1..239fbc2 100644
--- a/src/sdl2_compat.h
+++ b/src/sdl2_compat.h
@@ -321,4 +321,11 @@ typedef struct SDL_WindowShapeMode {
     SDL_WindowShapeParams parameters;
 } SDL_WindowShapeMode;
 
+
+typedef struct SDL2_Vertex {
+    SDL_FPoint position;
+    SDL_Color  color;
+    SDL_FPoint tex_coord;
+} SDL2_Vertex;
+
 #endif /* sdl2_compat_h */
diff --git a/src/sdl2_protos.h b/src/sdl2_protos.h
index 01f62d3..6a75a3b 100644
--- a/src/sdl2_protos.h
+++ b/src/sdl2_protos.h
@@ -870,7 +870,7 @@ SDL2_PROTO(int,JoystickSendEffect,(SDL_Joystick *a, const void *b, int c))
 SDL2_PROTO(float,GameControllerGetSensorDataRate,(SDL_GameController *a, SDL_SensorType b))
 SDL2_PROTO(int,SetTextureUserData,(SDL_Texture *a, void *b))
 SDL2_PROTO(void*,GetTextureUserData,(SDL_Texture *a))
-SDL2_PROTO(int,RenderGeometry,(SDL_Renderer *a, SDL_Texture *b, const SDL_Vertex *c, int d, const int *e, int f))
+SDL2_PROTO(int,RenderGeometry,(SDL_Renderer *a, SDL_Texture *b, const SDL2_Vertex *c, int d, const int *e, int f))
 SDL2_PROTO(int,RenderGeometryRaw,(SDL_Renderer *a, SDL_Texture *b, const float *c, int d, const SDL_Color *e, int f, const float *g, int h, int i, const void *j, int k, int l))
 SDL2_PROTO(int,RenderSetVSync,(SDL_Renderer *a, int b))
 SDL2_PROTO(int,asprintf,(char **a, SDL_PRINTF_FORMAT_STRING const char *b, ...))
diff --git a/src/sdl3_syms.h b/src/sdl3_syms.h
index 5fe0f52..f88ec3d 100644
--- a/src/sdl3_syms.h
+++ b/src/sdl3_syms.h
@@ -705,8 +705,7 @@ SDL3_SYM_PASSTHROUGH(int,FlashWindow,(SDL_Window *a, SDL_FlashOperation b),(a,b)
 SDL3_SYM_RENAMED(int,GameControllerSendEffect,SendGamepadEffect,(SDL_GameController *a, const void *b, int c),(a,b,c),return)
 SDL3_SYM_RENAMED(int,JoystickSendEffect,SendJoystickEffect,(SDL_Joystick *a, const void *b, int c),(a,b,c),return)
 SDL3_SYM_RENAMED(float,GameControllerGetSensorDataRate,GetGamepadSensorDataRate,(SDL_GameController *a, SDL_SensorType b),(a,b),return)
-SDL3_SYM(int,RenderGeometry,(SDL_Renderer *a, SDL_Texture *b, const SDL_Vertex *c, int d, const int *e, int f),(a,b,c,d,e,f),return)
-SDL3_SYM(int,RenderGeometryRaw,(SDL_Renderer *a, SDL_Texture *b, const float *c, int d, const SDL_Color *e, int f, const float *g, int h, int i, const void *j, int k, int l),(a,b,c,d,e,f,g,h,i,j,k,l),return)
+SDL3_SYM(int,RenderGeometryRaw,(SDL_Renderer *a, SDL_Texture *b, const float *c, int d, const SDL_FColor *e, int f, const float *g, int h, int i, const void *j, int k, int l),(a,b,c,d,e,f,g,h,i,j,k,l),return)
 SDL3_SYM_RENAMED(int,RenderSetVSync,SetRenderVSync,(SDL_Renderer *a, int b),(a,b),return)
 SDL3_SYM_VARARGS(int,asprintf,(char **a, SDL_PRINTF_FORMAT_STRING const char *b, ...))
 SDL3_SYM_PASSTHROUGH(int,vasprintf,(char **a, const char *b, va_list c),(a,b,c),return)
```

**Comment by slouken** [comment]
> testgeometry.c needs the following, no?

Yep, thanks!

**Comment by sezero** [comment]
> > > sdl2-compat side need updating too.
> > 
> > 
> > I tried this patch (inlined below) but testgeometry doesn't look exactly the same. [compat.patch.txt](https://github.com/libsdl-org/SDL/files/14107950/compat.patch.txt) I am missing something obvious? Will leave to you..
> 
> That looks like it ought to work. I'll check it out, thanks.

OK. (FWIW, it _works_ technically, yes, but all the colors are _lighter_ compared to normal SDL2)

**Comment by slouken** [comment]
Oh, go ahead and check that in then. I need to fix up the colors in SDL3. The testgeometry in main probably has the same problem.

**Comment by sezero** [comment]
Pushed the main patch as https://github.com/libsdl-org/sdl2-compat/commit/d5c09b873cbee7a7ee78282ac76852052b998689

Pushed a _just-in-case_ follow-up safeguard patch as https://github.com/libsdl-org/sdl2-compat/commit/98aa2fd7416a07916ffbdec9b8c142139fb564fe

> Question: Should I keep temporary allocation as `stack_alloc` or should I use that `small_alloc` thing ?

This question still stands

**Comment by slouken** [comment]
> Question: Should I keep temporary allocation as `stack_alloc` or should I use that `small_alloc` thing ?

It should use the small alloc thing.

**Comment by slouken** [comment]
Okay, I double checked and the lighter colors you're seeing are expected, and caused by interpolating in linear space instead of sRGB space. We're still thinking about whether defaulting to linear space is a good idea, but that behavior is expected right now.

**Comment by sezero** [comment]
> > Question: Should I keep temporary allocation as `stack_alloc` or should I use that `small_alloc` thing ?
> 
> It should use the small alloc thing.

Done: check with the latest git version.

> Okay, I double checked and the lighter colors you're seeing are expected, and caused by interpolating in linear space instead of sRGB space. We're still thinking about whether defaulting to linear space is a good idea, but that behavior is expected right now.

OK.

**Comment by ocornut** [comment]
Hello,

Would it be possible for SDL to maintain a variant of SDL_RenderGeometryRaw() that took packed colors and didn’t incur an extra transformation?

It seems like this change would have meaningful effect on low-end machines (eg I noticed a PS2 rendered being mentioned) where some users now may be inclined to convert their packed colors to float4 and then the SDL backend is going to be doing the exact opposite.

Does SDL3 provide an alternative way (akin to vertex layout configuration) to allow rendering with packed colors?

(ref this is breaking dear imgui backend, i am not thrilled about converting 30k vertices every frame when GPU can do it supposedly? https://github.com/ocornut/imgui/issues/7293#issuecomment-1925656970)

**Comment by slouken** [comment]
Yeah, I've been mulling this over myself. Can you enter a new issue for tracking this?

**Pull request merged**